### PR TITLE
Refactor fx Start/Stop usage

### PR DESCRIPTION
**Pull request by dnr**
**What changed?**
- Move blocking start outside of fx lifecycle hook
- Propagate outer app Start/Stop context to inner Apps (currently just Background)
- Refactor StopService into proper method
- Other small cleanups

This leaves the shutdown timeout in the inner Apps: if we set a single timeout on the outer one, fx itself might abort shutdown and not give the inner ones time to finish, so I think it's more robust this way.

**Why?**
Simplify and correct usage of fx

**How did you test it?**
Local testing, integration tests

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: temporal/fx.go [hunk 1]
```diff
@@ -549,13 +548,7 @@ func WorkerServiceProvider(
 
 	if _, ok := params.ServiceNames[serviceName]; !ok {
 		params.Logger.Info("Service is not requested, skipping initialization.", tag.Service(serviceName))
-		return ServicesGroupOut{
-			Services: &ServicesMetadata{
-				App:           fx.New(fx.NopLogger),
-				ServiceName:   serviceName,
-				ServiceStopFn: func() {},
-			},
-		}, nil
```
Comment (MichaelSnowden) [inline temporal/fx.go:558]: This is a great change, but can we move it to a separate PR?

**Review by MichaelSnowden — CHANGES_REQUESTED** [review]
This definitely looks correct to me at a glance, but I think the unrelated parts should be split into separate PRs, to make it harder to miss any issues

File: temporal/fx.go [hunk 2]
```diff
@@ -88,19 +84,23 @@ type (
...
+		app         *fx.App
+		serviceName primitives.ServiceName
+		logger      log.Logger
+		stopChan    chan struct{}
 	}
 
 	ServerFx struct {
-		app *fx.App
+		app           *fx.App
+		blockingStart blockingStartParams
```
[the remark this answers is not in the document]
Comment (dnr) [reply temporal/fx.go:95]: sure, that's a nice name

File: temporal/fx.go [hunk 3]
```diff
@@ -119,21 +119,21 @@ type (
 		AudienceGetter         authorization.JWTAudienceMapper
 
 		// below are things that could be over write by server options or may have default if not supplied by serverOptions.
-		Logger                  log.Logger
-		ClientFactoryProvider   client.FactoryProvider
-		DynamicConfigClient     dynamicconfig.Client
-		DynamicConfigCollection *dynamicconfig.Collection
```
[the remark this answers is not in the document]
Comment (dnr) [reply temporal/fx.go:125]: alright

File: temporal/fx.go [hunk 4]
```diff
@@ -148,11 +148,14 @@ func NewServerFx(opts ...ServerOption) (*ServerFx, error) {
 		fx.Provide(ApplyClusterMetadataConfigProvider),
 		fx.Invoke(ServerLifetimeHooks),
 		FxLogAdapter,
+
+		fx.Populate(&s.blockingStart),
+		fx.Populate(&s.logger),
```
[the remark this answers is not in the document]
Comment (dnr) [reply temporal/fx.go:153]: The `ServerFx` object is constructed right here. `NewServerFxImpl` provides a `*ServerImpl`, not a `ServerFx` (the names in here are a twisty maze…)

`*ServerFx` implements `Server`, but `*ServerImpl` doesn't. It previously did, but the point was adding contexts, so it can't anymore.

I can rename stuff to make it less confusing but tbh I'm not even sure where to start :sweat_smile:  Maybe renaming `NewServerFxImpl` to `NewServerImpl`? Or `ServerImpl` to …? And then `ServerFx` to `ServerImpl`?

File: temporal/fx.go (hunk 1 above)
Comment (dnr) [reply temporal/fx.go:558]: I split out the Collection one (already in a separate commit) but I'd rather not spend time on splitting the rest, they're fairly intertwined

**Review by MichaelSnowden — APPROVED** [review]

File: temporal/fx.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply temporal/fx.go]: ...let's figure this out later lol
